core/trino-main/src/test/java/io/trino/operator/PageTestUtils.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.block.BlockAssertions;
+import io.trino.spi.Page;
+import io.trino.spi.PageBuilder;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.type.Type;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.google.common.base.Verify.verify;
+import static io.trino.block.BlockAssertions.Encoding.DICTIONARY;
+import static io.trino.block.BlockAssertions.Encoding.RUN_LENGTH;
+import static io.trino.block.BlockAssertions.createAllNullsBlock;
+import static io.trino.block.BlockAssertions.createRandomBlockForType;
+import static io.trino.block.BlockAssertions.createRandomLongsBlock;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static java.lang.String.format;
+
+public class PageTestUtils
+{
+    public static Page createPageWithRandomData(List<Type> types, int positionCount, float primitiveNullRate, float nestedNullRate)
+    {
+        return createPageWithRandomData(types, positionCount, true, false, primitiveNullRate, nestedNullRate, ImmutableList.of());
+    }
+
+    public static Page createDictionaryPageWithRandomData(List<Type> types, int positionCount, float primitiveNullRate, float nestedNullRate)
+    {
+        return createPageWithRandomData(types, positionCount, true, false, primitiveNullRate, nestedNullRate, ImmutableList.of(DICTIONARY));
+    }
+
+    public static Page createRlePageWithRandomData(List<Type> types, int positionCount, float primitiveNullRate, float nestedNullRate)
+    {
+        return createPageWithRandomData(types, positionCount, true, false, primitiveNullRate, nestedNullRate, ImmutableList.of(RUN_LENGTH));
+    }
+
+    public static Page createPageWithRandomData(
+            List<Type> types,
+            int positionCount,
+            boolean addPreComputedHashBlock,
+            boolean addNullBlock,
+            float primitiveNullRate,
+            float nestedNullRate,
+            List<BlockAssertions.Encoding> wrappings)
+    {
+        int channelCount = types.size();
+
+        List<Block> blocks = new ArrayList<>();
+
+        if (addPreComputedHashBlock) {
```

Comment:
this is wrong. you don't provide any hash channels. Hence hash is not really precomputed on any column (see how we do it in `io.trino.RowPagesBuilder#pagesWithHash`).